File: mathematics/fundamentals/mutual-recurrences.py
# Mathematics > Fundamentals > Mutual Recurrences
# Compute terms in a mutual recurrence.
#
# https://www.hackerrank.com/challenges/mutual-recurrences/problem
# https://www.hackerrank.com/contests/infinitum14/challenges/mutual-recurrences
# challenge id: 15898
#

import sys

MOD = 1000000000

#------------------------------------------------------------------------------

a, b, c, d, e, f, g, h = 0, 0, 0, 0, 0, 0, 0, 0
cache_x = None
cache_y = None


def x(n):
    if n < 0:
        return 1
    v =  cache_x[n]
    if v == 0:
        v = x(n - a) + y(n - b) + y(n - c) + n * d ** n
        cache_x[n] = v
    return v


def y(n):
    if n < 0:
        return 1
    v = cache_y[n]
    if v == 0:
        v = y(n - e) + x(n - f) + x(n - g) + n * h ** n
        cache_y[n] = v
    return v

#------------------------------------------------------------------------------

class Matrice:
    def __init__(self, n, v=None):
        self.n = n
        self.v = [0] * (n * n)
        if v is not None:
            if isinstance(v, int) and v == 1:
                for i in range(n):
                    self.v[i + n * i] = 1
            elif isinstance(v, list) and len(v) == n * n:
                self.v = v
            else:
                raise ValueError

    def __setitem__(self, xy, v):
        self.v[xy[0] + self.n * xy[1]] = v

    def __getitem__(self, xy):
        return self.v[xy[0] + self.n * xy[1]]

    def __call__(self, x, y):
        return self.v[x + self.n * y]

    def __str__(self):
        n = self.n
        m = [max(len(str(self.v[x + n * y])) for y in range(n))
             for x in range(n)]
        s = ''
        for i in range(0, n * n, n):
            s += '[' if i == 0 else ' '
            s += '  '.join(str(self.v[j]).rjust(m[j - i]) for j in range(i, i + n))
            s += ']' if i == n * n - n else '\n'
        return s

    def __imod__(self, m):
        """ A %= m """
        for i in range(self.n * self.n):
            self.v[i] %= m
        return self

    def tr(self):
        n = self.n
        Ap = Matrice(n)
        for x in range(n):
            for y in range(n):
                Ap.v[x + n * y] = self.v[y + n * x]
        return Ap


class Vecteur:
    def __init__(self, n, v=None):
        self.n = n
        self.v = [0] * n
        if v is not None:
            if isinstance(v, list) and len(v) == n:
                self.v = v
            else:
                raise ValueError

    def __setitem__(self, x, v):
        self.v[x] = v

    def __getitem__(self, x):
        return self.v[x]

    def __call__(self, x):
        return self.v[x]

    def __str__(self):
        n = self.n
        s = '[' + ' '.join(map(str, self.v)) + ']'
        return s

    def __imod__(self, m):
        """ V %= m """
        for i in range(self.n):
            self.v[i] %= m
        return self


def multm(A, B):
    """ produit matriciel: A * B """
    assert A.n == B.n
    n = A.n
    r = Matrice(n)
    for x in range(n):
        for y in range(n):
            r[(x, y)] = sum(A(i, y) * B(x, i) for i in range(n)) % MOD
    return r


def multv(A, V):
    """ produit matrice/vecteur: A * V """
    assert A.n == V.n
    n = A.n
    r = Vecteur(n)
    for y in range(n):
        r[y] = sum(A(i, y) * V(i) for i in range(n)) % MOD
    return r


def power(M, k):
    """ fast exponentiation M^k """
    P = Matrice(M.n, 1)
    if k == 0:
        return P
    if k == 1:
        return M
    while k != 0:
        if k % 2 == 1:
            P = multm(P, M)
        M = multm(M, M)
        k //= 2
    return P


#------------------------------------------------------------------------------

def fibonacci_easy(a, b, n):
    A = Matrice(2, [1, 1, 1, 0])
    An = power(A, n)
    F0 = Vecteur(2, [b, a])
    Fn = multv(An, F0)
    print(Fn[1])


def solve(a, b, c, d, e, f, g, h, n):

    A = Matrice(24)

    A[(     a - 1, 0)] += 1         # x(n) =  x(n - a)
    A[(10 + b - 1, 0)] += 1         #                  + y(n - b)
    A[(10 + c - 1, 0)] += 1         #                             + y(n - c)
    A[(        22, 0)] += 1         #                                        + n * d ** n

    A[(10 + e - 1, 10)] += 1        # y(n) = y(n - e)
    A[(     f - 1, 10)] += 1        #                 + x(n - f)
    A[(     g - 1, 10)] += 1        #                             + x(n - g)
    A[(        20, 10)] += 1        #                                        + n * h ** n

    for i in range(1, 10):
        A[(i - 1, i)] += 1
        A[(10 + i - 1, 10 + i)] += 1

    A[(20, 20)] = h
    A[(21, 20)] = h
    A[(21, 21)] = h

    A[(22, 22)] = d
    A[(23, 22)] = d
    A[(23, 23)] = d

    V = Vecteur(24, [1] * 24)
    V[20] = 0
    V[22] = 0

    # print(A, file=sys.stderr)

    A = power(A, n + 1)
    r = multv(A, V)
    print(r[0], r[10])


if __name__ == '__main__':
    for _ in range(int(input())):
        a, b, c, d, e, f, g, h, n = map(int, input().split())

        """ naif
        cache_x = [0] * (n + 1)
        cache_y = [0] * (n + 1)
        print(x(n) % MOD, y(n) % MOD)
        """

        solve(a, b, c, d, e, f, g, h, n)

File: mathematics/fundamentals/mutual-recurrences.cpp
// Mathematics > Fundamentals > Mutual Recurrences
// Compute terms in a mutual recurrence.
//
// https://www.hackerrank.com/challenges/mutual-recurrences/problem
// https://www.hackerrank.com/contests/infinitum14/challenges/mutual-recurrences
// challenge id: 15898
//

#include <bits/stdc++.h>
using namespace std;

#pragma GCC diagnostic ignored "-Wsign-conversion"

constexpr long MOD = 1e9;

class Vecteur
{
    vector<long> a;
    size_t n;

  public:
    Vecteur(size_t n)
    {
        this->n = n;
        a.resize(n, 0L);
    }

    size_t dim() const
    {
        return n;
    }

    long operator()(size_t x) const
    {
        assert(x < n);
        return a[x];
    }

    long &operator()(size_t x)
    {
        assert(x < n);
        return a[x];
    }
};

class Matrice
{
    vector<long> a;
    size_t n;

  public:
    Matrice(size_t n)
    {
        this->n = n;
        a.resize(n * n, 0L);
    }

    static Matrice I(size_t n)
    {
        Matrice M(n);
        for (size_t i = 0; i < n; ++i)
            M(i, i) = 1;
        return M;
    }

    long operator()(size_t x, size_t y) const
    {
        assert(x < n && y < n);
        return a[x + y * n];
    }

    long &operator()(size_t x, size_t y)
    {
        assert(x < n && y < n);
        return a[x + y * n];
    }

    Matrice &operator%=(long m)
    {
        for (auto &&i : a)
            i %= m;
        return *this;
    }

    Matrice &operator*=(const Matrice &B)
    {
        assert(n == B.n);

        Matrice &A = *this;
        Matrice C(n);

        for (size_t x = 0; x < n; ++x)
        {
            for (size_t y = 0; y < n; ++y)
            {
                long s = 0;
                for (size_t i = 0; i < n; ++i)
                    s += (A(i, y) * B(x, i)) % MOD;
                C(x, y) = s % MOD;
            }
        }
        a = C.a;
        return *this;
    }

    Vecteur operator*(const Vecteur &V) const
    {
        assert(n == V.dim());
        const Matrice &A = *this;
        Vecteur C(n);

        for (size_t y = 0; y < n; ++y)
        {
            long s = 0;
            for (size_t i = 0; i < n; ++i)
                s += (A(i, y) * V(i)) % MOD;
            C(y) = s % MOD;
        }

        return C;
    }

    Matrice &operator^=(unsigned long k)
    {
        if (k == 1)
            return *this;

        Matrice &M = *this;
        Matrice P = Matrice::I(n);

        while (k != 0)
        {
            if (k % 2 == 1)
                P *= M;
            M *= M;
            k /= 2;
        }

        a = P.a;
        return *this;
    }

    void print() const
    {
        const Matrice &A = *this;

        for (size_t y = 0; y < n; ++y)
        {
            cout << ((y == 0) ? '[' : ' ');
            for (size_t x = 0; x < n; ++x)
            {
                cout << A(x, y);
                if (x == n - 1)
                {
                    if (y == n - 1)
                        cout << "]\n";
                    else
                        cout << "\n";
                }
                else
                {
                    cout << " ";
                }
            }
        }
    }
};

long fibonacci_easy(long a, long b, unsigned long n)
{
    Matrice A(2);

    A(0, 0) = 1;
    A(0, 1) = 1;
    A(1, 0) = 1;
    A(1, 1) = 0;

    A ^= n;

    Vecteur F(2);
    F(0) = b;
    F(1) = a;

    F = A * F;

    return F(1);
}

// Complete the solve function below.
void solve(int a, int b, int c, int d, int e, int f, int g, int h, unsigned long n)
{
    Matrice A(20 + 4);

    A(a - 1, 0) += 1;      // x(n) =  x(n - a)
    A(10 + b - 1, 0) += 1; //                  + y(n - b)
    A(10 + c - 1, 0) += 1; //                             + y(n - c)
    A(20 + 2, 0) += 1;     //                                        + n * d ** n

    A(10 + e - 1, 10) += 1; // y(n) = y(n - e)
    A(f - 1, 10) += 1;      //                 + x(n - f)
    A(g - 1, 10) += 1;      //                             + x(n - g)
    A(20, 10) += 1;         //                                        + n * h ** n

    for (size_t i = 1; i < 10; ++i)
    {
        A(i - 1, i) += 1;
        A(10 + i - 1, 10 + i) += 1;
    }

    A(20, 20) = h;
    A(20 + 1, 20) = h;
    A(20 + 1, 20 + 1) = h;

    A(20 + 2, 20 + 2) = d;
    A(20 + 3, 20 + 2) = d;
    A(20 + 3, 20 + 3) = d;

    //A.print();

    Vecteur F(20 + 4);
    for (size_t i = 0; i < 20 + 4; ++i)
        F(i) = 1;
    F(20) = 0;
    F(20 + 2) = 0;

    A ^= (n + 1);
    F = A * F;
    cout << F(0) << " " << F(10) << endl;
}

int main()
{
    // cerr << fibonacci_easy(2, 4, 9) << endl;        // [55, 34, 34, 21]  178

    int t;
    cin >> t;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    while (t--)
    {
        int a, b, c, d, e, f, g, h;
        unsigned long n;
        cin >> a >> b >> c >> d >> e >> f >> g >> h >> n;
        solve(a, b, c, d, e, f, g, h, n);
    }

    return 0;
}
